File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        # approach: binary search on range [1, max(piles)]
        
        result = float('inf')
        l = 1
        r = max(piles)
        while l <= r:
            m = (l + r) // 2
            
            hours = 0
            for pile in piles:
                hours += math.ceil(pile / m)

            if hours > h:
                l = m + 1
            else:
                r = m - 1
                
                # potential new min k
                result = min(result, m)
        
        return result

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        // approach: binary search on piles.min, piles.max
        
        // get max of piles
        int max = 0;
    
        for (int pile : piles) {
            max = Math.max(max, pile);
        }
        
        // binary search on [1, max]
        int l = 1;
        int r = max;
        int k = max;
        int result = max;
        
        while (l <= r) {
            k = (l + r) / 2;
            int hours = 0;
            
            for (int pile : piles) {
                hours += Math.ceil((double) pile / k);
            }
            
            if (hours > h) {
                l = k + 1;
            } else {
                // potential new min
                result = Math.min(result, k);
                r = k - 1;
            }
        }
        
        return result;
    }
}